contracts/Dispenser.sol

```diff
@@ -113,4 +237,373 @@ contract Dispenser is IErrorsTokenomics {
+        // Shall not claim in the same epoch
+        if (eCounter == lastClaimedEpoch) {
+            revert();
+        }
+
+        // TODO: Register lastClaimedEpoch for the first time? Here or via VoteWeighting?
+        // TODO: Define a maximum number of epochs to claim for
+        for (uint256 j = lastClaimedEpoch; j < eCounter; ++j) {
+            // TODO: optimize not to read several times in a row same epoch info
+            // Get service staking info
+            ITokenomicsInfo.ServiceStakingPoint memory serviceStakingPoint =
+                ITokenomicsInfo(tokenomics).mapEpochServiceStakingPoints(j);
+
+            ITokenomicsInfo.EpochPoint memory ep = ITokenomicsInfo(tokenomics).mapEpochTokenomics(j);
+            uint256 endTime = ep.endTime;
+
+            // Get the staking weight for each epoch and the total weight
+            // TODO math from where we need to get the weight - endTime or endTime + WEEK, seems like the latter because values are written for the nextTime
+            (uint256 stakingWeight, uint256 totalWeighSum) =
+                IVoteWeighting(voteWeighting).nomineeRelativeWeight(target, chainId, endTime);
+
+            // Adjust the inflation by the maximum amount of veOLAS voted for service staking contracts
+            uint256 availableStakingAmount = serviceStakingPoint.serviceStakingAmount;
+            if (availableStakingAmount > totalWeighSum) {
+                availableStakingAmount = totalWeighSum;
+            }
+
+            // Compare the staking weight
+            uint256 stakingAmount;
+            uint256 returnAmount;
+            if (stakingWeight < serviceStakingPoint.minServiceStakingWeight) {
+                // If vote weighting staking weight is lower than the defined threshold - return the staking amount
+                returnAmount = (availableStakingAmount * stakingWeight) / 1e18;
+                totalAmountReturn += returnAmount;
+            } else {
+                // Otherwise, allocate staking amount to corresponding contracts
+                stakingAmount = (availableStakingAmount * stakingWeight) / 1e18;
+                if (stakingAmount > serviceStakingPoint.maxServiceStakingAmount) {
+                    // Adjust the return amount
+                    returnAmount = stakingAmount - serviceStakingPoint.maxServiceStakingAmount;
+                    totalAmountReturn += returnAmount;
+                    // Adjust the staking amount
+                    stakingAmount = serviceStakingPoint.maxServiceStakingAmount;
+                }
+                totalStakingAmount += stakingAmount;
+            }
+
+            // Adjust remaining service staking amounts
+            uint256 remainingServiceStakingAmount = mapEpochRemainingServiceStakingAmounts[j];
+            // Combine staking and return amounts as they both need to be subtracted from the remaining amount
+            stakingAmount += returnAmount;
```

Comment:
add an event saying that the target contract can receive at most maxServiceStakingAmount OLAS per epoch